### PR TITLE
Add `globby.stream`

### LINKED ISSUE #83: Async-iteration support
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/globby/issues/83)

It would be great to see results as they are found. By implementing JavaScript async iteration support, we could be able to have results stream in as they are located.

Filing as a feature request here, because this is my favorite glob library. 🤩 Thanks!!!

> There is a $80.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/globby/issues/83).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/globby/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by futpib**
Fixes #83

This adds support for async iteration by adding `globby.stream` which returns a `ReadableStream`. 

Node 10 introduced experimental [async iterator support to `ReadableStream`](https://nodejs.org/api/stream.html#stream_readable_symbol_asynciterator) so if supported by one's node version, one can use `globby.stream` in a `for-await-of` loop as witnessed by this test https://github.com/futpib/globby/blob/958e81505b1baa7d73e29dbe3468123032cce9cf/test.js#L85-L92

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -120,24 +128,36 @@ module.exports.default = globby;
...
 	return tasks.reduce(
 		(matches, task) => arrayUnion(matches, fastGlob.sync(task.pattern, task.options)),
 		[]
 	).filter(p => !filter(p));
 };
 
+module.exports.stream = (patterns, options) => {
+	const globTasks = generateGlobTasks(patterns, options);
+
+	const tasks = globTasks.reduce((tasks, task) => {
```
Comment (alexzherdev) [inline index.js]: Is this equivalent to a `.map` and would that be more clear?

Comment (futpib) [reply index.js]: This is equivalent to a `.flatMap`.

File: stream-utils.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+'use strict';
```
Comment (sindresorhus) [inline stream-utils.js]: This file must be added to https://github.com/sindresorhus/globby/blob/04d51bf1b85ad9c7ec968126b26c6d6b7eae92be/package.json#L21

**Comment by sindresorhus** [comment]
You also need to update https://github.com/sindresorhus/globby/blob/master/index.d.ts

**Comment by sindresorhus** [comment]
Can you fix the merge conflicts?

File: test.js [hunk 3]
```diff
@@ -221,6 +277,11 @@ test('respects gitignore option false - sync', t => {
 	t.true(actual.indexOf('node_modules') > -1);
 });
 
+test('respects gitignore option false - stream', async t => {
+	const actual = await getStream.array(globby.stream('*', {gitignore: false, onlyFiles: false}));
+	t.true(actual.indexOf('node_modules') > -1);
```
[the remark this answers is not in the document]
Comment (futpib) [reply test.js]: Ah, `indexOf` was consistent with the rest of the file at the time of writing.

**Comment by futpib** [comment]
I have addressed all the review comments :ok_hand:

File: readme.md [hunk 4]
```diff
@@ -93,6 +93,18 @@ Respect ignore patterns in `.gitignore` files that apply to the globbed files.
 
 Returns `string[]` of matching paths.
 
+### globby.stream(patterns, options?)
+
+Returns a [`stream.Readable`](https://nodejs.org/api/stream.html#stream_readable_streams) of matching paths.
+
+Since Node.js 10, [readable streams are iterable](https://nodejs.org/api/stream.html#stream_readable_symbol_asynciterator), so you can loop over glob matches in a [`for await...of` loop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) like this:
```
Comment (sindresorhus) [inline readme.md]: This an the below example should be in the TS definition too.

Comment (futpib) [reply readme.md]: Do you mean in the docs for the `stream` method around here
https://github.com/sindresorhus/globby/blob/6618d820fa6bd0580dae6d675984dbd4cfe5ab4a/index.d.ts#L117

, or in the `index.test-d.ts`? I've already added a loop like this there:
https://github.com/sindresorhus/globby/blob/6618d820fa6bd0580dae6d675984dbd4cfe5ab4a/index.test-d.ts#L68

Comment (sindresorhus) [reply readme.md]: > Do you mean in the docs for the stream method around here

Yes

**Comment by sindresorhus** [comment]
Thanks 🙌

You need to submit the PR URL here to claim the bounty: https://issuehunt.io/r/sindresorhus/globby/issues/83

**Pull request merged**